hathora/hathora-unreal-sdk
# Implement support for ICMP pinging

This PR supersedes #6 by replacing Websocket pinging with ICMP echos. Unreal's built in `FIcmp::IcmpEcho` function handles DNS resolution timing separate from the actual ICMP echo, so the logic ends up being very simple without extra overrides for accurate results.

This PR also implements various refactoring improvements to the project/plugin:
- `FHathoraSDK` was renamed to `FHathoraSDKModule`; I did this because the module itself isn't useful for Blueprint support and I wanted to use the `HathoraSDK` name for a generic library
- `UHathoraPing` was renamed to `UHathoraSDK` and it was reparented with `UBlueprintFunctionLibrary`. Blueprint Function Libraries can still be used by C++, but these classes are special wrappers to help expose functions like these API calls to blueprints.  I also didn't find much purpose to prematurely separate out the ping SDK function (`GetRegionalPings`) into it's own class, so I figured we should just put these in a common `UHathoraSDK` class that can cover most uses until it makes sense to break things out.
- Switching to a `UBlueprintFunctionLibrary` implies that the class is merely a scoping construct for static functions; these classes don't have state, and up until this point, `UHathoraPing` didn't use state other than the config (which was changed to support this, see the next bullet). A call to the sdk is simply `UHathoraSDK::GetRegionalPings(OnComplete)`
- `UHathoraSdkConfig` was renamed to `UHathoraSDKConfig` for consistency
- `UHathoraSDKConfig` now inherits from `UDeveloperSettings` which allows us to have these parameters to be located in UE's Edit > Project Settings > Plugins > Hathora SDK. It also prevents needing a separate `UHathoraSDKConfig.cpp` file since the values are fetched using the CDO. Unreal's `GetDefault<UHathoraSDKConfig>` will read whatever is in the Project Settings and/or in `DefaultGame.ini` or `Game.ini` to read in the values. It all kind of magically works under the hood. This prevents us from needing to have some instanced config object, and is closer to how configuration works in Unreal.
- I changed the logic of `GetRegionalPings` to use a Dynamic Delegate instead of a normal Delegate to be able to support Blueprint callbacks; Dynamic Delegates only work in UObjects, so I needed to change the `SDemoMenuWidget` in the project to a `UWidget`; I opted for replacing it with a complete `UUserWidget`.
- Since I was already modifying the demo widget, I added some more visual indicators of the pings instead of just logging it to the output.
- I changed the project to default to 5.3 since that's the latest Unreal version

![image](https://github.com/hathora/hathora-unreal-sdk/assets/549323/ec72e5d5-249f-47e8-bac3-995b5d69e78b)

**Simplest testing steps:**
1. Open the project in Unreal 5.3. If you want to open it in 5.2, you should be able to right click on the uproject file in File Explorer, click "Show more options" (if in Win 11), and then "Switch Unreal Engine version...". You may have to open the generated solution (depending if you're using VS, Rider, or VSCode) and recompile the project manually.
2. Once Unreal is open, click on Edit > Project Settings > find the Plugins section on the left > click on Hathora SDK.
3. Edit the Base URL to the correct Hathora development API URL that has ICMP support; no need to click Save or Export, it's set now
4. Press the Play button
5. You should see a black screen with a button that says Get Pings; press the button
6. You should see a list of 5 regions and their pings like the above image

**Code review suggestion:**
Due to the renaming, just looking at File Changes in GitHub may be a bit convoluted. I tried to separate the commits out to logical steps of the process, and it may be easier to follow the changes by reviewing the commits themselves.
